src/main/java/org/broadinstitute/hellbender/utils/hmm/segmentation/HiddenMarkovModelSegmentProcessor.java

```diff
@@ -0,0 +1,652 @@
+import htsjdk.samtools.util.Locatable;
+import org.apache.commons.lang3.tuple.ImmutablePair;
+import org.apache.commons.lang3.tuple.Pair;
+import org.apache.commons.math3.util.FastMath;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.broadinstitute.hellbender.exceptions.GATKException;
+import org.broadinstitute.hellbender.exceptions.UserException;
+import org.broadinstitute.hellbender.tools.exome.HashedListTargetCollection;
+import org.broadinstitute.hellbender.tools.exome.Target;
+import org.broadinstitute.hellbender.tools.exome.TargetCollection;
+import org.broadinstitute.hellbender.tools.exome.sexgenotyper.SexGenotypeData;
+import org.broadinstitute.hellbender.utils.IndexRange;
+import org.broadinstitute.hellbender.utils.IntervalUtils;
+import org.broadinstitute.hellbender.utils.QualityUtils;
+import org.broadinstitute.hellbender.utils.Utils;
+import org.broadinstitute.hellbender.utils.hmm.ForwardBackwardAlgorithm;
+import org.broadinstitute.hellbender.utils.hmm.interfaces.CallStringProducer;
+import org.broadinstitute.hellbender.utils.hmm.interfaces.ScalarProducer;
+
+import javax.annotation.Nonnull;
+import java.io.IOException;
+import java.util.*;
+import java.util.function.BiFunction;
+import java.util.stream.Collectors;
+import java.util.stream.IntStream;
+
+/**
+ * TODO github/gatk-protected issue #855 -- this is a last-minute fork of {@link HiddenMarkovModelPostProcessor}.
+ * It just does segmentation (no VCF creation)
+ *
+ * @author Mehrtash Babadi &lt;mehrtash@broadinstitute.org&gt;
+ */
+public class HiddenMarkovModelSegmentProcessor<D, S extends CallStringProducer & ScalarProducer, T extends Target> {
+
+    private final Logger logger = LogManager.getLogger(HiddenMarkovModelSegmentProcessor.class);
+
+    /**
+     * Maximum tolerated log probability value. Values between this constant and 0.0 as considered as a 0.0.
+     * Values above this threshold are considered to indicate a log probability calculation problem that is
+     * worth to be reported as an error or warning.
+     */
+    public static final double MAX_LOG_PROB = 1e-3;
+
+    /**
+     * Maximum reportable output quality score; higher quality scores will
+     * capped to this value.
+     */
+    public static final double MAX_QUAL_SCORE = 199.99999999999;
+
+    /**
+     * Precision of Phred-scale scores
+     */
+    public static final double PHRED_SCORE_PRECISION = 0.1;
+
+    /**
+     * Calculating variance posterior on long segments is time consuming; we assume hidden states become
+     * independent if their separation is larger than the following value
+     */
+    public static final double INDEPENDENT_TARGETS_SEPARATION_THRESHOLD = 0;
+
+    /**
+     * Threshold used to determine best way to calculate log(1- exp(a))
+     * based on https://cran.r-project.org/web/packages/Rmpfr/vignettes/log1mexp-note.pdf
+     */
+    private static final double LN_1_M_EXP_THRESHOLD = - Math.log(2);
```

Comment:
Refactored `logProbComplement`, `roundPhred`, and `logProbToPhredScore` to `GATKProtectedMathUtils`. Regarding specific unit tests to these methods -- I think it is unnecessary (they are excessively simple).

Some unit tests for `HiddenMarkovModelPostProcessor` and `HiddenMarkovModelSegmentProcessor` will be useful. The latter is a quick fork of the former without VCF creation functionality. I will add some test placeholders for the time being.